### DISCUSSION TITLE
k8s helm deployment with microk8s, calico, nginx ingress and open relay

### DISCUSSION THREAD

**Discussion by huehnerhose** [Mailu/Mailu#3609] [User support] [answered]
Hi, 
I am banging my head against the desk with this:
My server is running k8s via microk8s, which comes with calico nicely configured. I am running nginx ingress controller and cert-manager and all this for handling certificates etc. All of this nicely runs with other workloads and also mailu is running "too good".
For deploying mailu I use the helm chart in version 2.1.1. 
Here the, i hope, relevant parts:
```
subnet: 10.1.164.0/24
front:
  hostPort:
    enabled: false
  externalService:
    enabled: false
ingress:
  realIpHeader: X-Forwarded-For
  realIpFrom: "10.1.164.32"
  enabled: true
```

Currently I am running an open relay :( My current understanding of the issue: since all incoming traffic goes via the nginx loadbalancer service, all request IPs are overridden with internal IPs. The realIpHeader / realIpFrom fields only matter for the admin UI / webmailer in the front pod. 
All smtp traffic is considered internal, since it originates from the same subnet, since it comes from the ingress. And since the HTTP-Header fields are this, *HTTP* header fields, there is no way filtering the SMTP traffic based on this. 

First: Is my understanding here right? Mainly about the realIpHeader/from and the front-pods filtering. 
Second: What now 😓 

Is there a way to make the front pod understand that this is still external traffic? 
Currently I am looking into creating a separate calico ippool and manually assigning ips to the pods, so that the "internal" IP of the ingress will be "external" to mailu. But this feels like overkill and quite cumbersome, since I will have to change the helm chart to allow for individual annotations for each pod. 

Any pointers, ideas and guidance highly appreciated!

**Comment by nextgens** [comment]
https://kubernetes.io/docs/reference/networking/service-protocols/#protocol-proxy-special

**Reply by huehnerhose** [reply]
Thanks for the answer. But I don't get it. The proxy protocol values are only used in the `mailu.proxyProtocolPorts` template. Which is not used anywhere in the helm chart. Further this template is only creating a list of ports when external services is enabled. Which is also not correct in my case, since I can't deploy another LoadBalancer. 

What am I missing?

**Reply by ghostwheel42** [reply, marked as answer]
You must not use any kind of address rewrite (NAT) for mail servers unless you can use "proxy protocol" to provide the real addresses. If you can't deploy this kind of load balancer then you cannot run mailu.
It may be possible to use a DSR setup to pass the real client addresses to mailu - I don't know if anyone did do this before. Maybe you can ask in the matrix channel.

**Reply by huehnerhose** [reply]
Thanks again. 
I was finally able to solve this. I got rid of calico. My microk8s now runs only metallb. Here I am able to share the IP by defining in the nginx-ingress and the mailu external service the service annotation: 
```metallb.universe.tf/allow-shared-ip: $IP```
See [here](https://metallb.universe.tf/usage/) for more details. Despite written in teh documentation both my loadbalancer services use externalNetworkPolicy: Local. But I had to explicitly set loadBalancerIP for both services. This now delivers the request IP properly to front-pod. 

So thanks!
